src/tests/integration/monitor-tabs.test.ts

```diff
@@ -0,0 +1,147 @@
+import { expect } from 'chai'
+import { createMocks } from './mocks'
+import { runBackground } from './steps/run-background'
+
+describe.only('monitor tabs', () => {
+  const mocks = createMocks()
+
+  runBackground(mocks, { alreadyAuthenticated: true })
+
+  describe('chrome.tabs.onCreated', () => {
+    it("adds the newly created tab to the store's state", () => {
+      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
+      const [callback] = mocks.chrome.tabs.onCreated.addListener.firstCall.args
+      callback({
+        id: 18,
+        windowId: 1,
+        active: false,
+        url: 'https://new.com/abc',
+      })
+      const state = mocks.getState()
+      const tab = state.tabs.get(18)!
+      expect(tab.id).to.equal(18)
+      expect(tab.windowId).to.equal(1)
+      expect(tab.domain).to.equal('new.com')
+      expect(tab.url).to.equal('https://new.com/abc')
+    })
+  })
+
+  describe('chrome.tabs.onRemoved', () => {
+    it("removes a tab from the store's state", () => {
+      const [callback] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
+      const toRemoveId = 11
+      callback(toRemoveId)
+      expect(mocks.getState().tabs.get(toRemoveId)).to.equal(undefined)
+    })
+  })
+
+  describe('chrome.tabs.onUpdated', () => {
+    it("updates a tab change to store's state", () => {
+      const [callbackOnUpdated] = mocks.chrome.tabs.onUpdated.addListener.firstCall.args
+
+      const changeInfo = {
+        id: 16,
+        windowId: 3,
+        active: false,
+        url: 'https://changeinfo.com/abc',
+      }
+
+      callbackOnUpdated(changeInfo.id, changeInfo)
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(changeInfo.id)!
+
+      expect(tab.id).to.equal(changeInfo.id)
+      expect(tab.windowId).to.equal(changeInfo.windowId)
+      expect(tab.domain).to.equal('changeinfo.com')
+      expect(tab.url).to.equal('https://changeinfo.com/abc')
+    })
+  })
+
+  describe('chrome.tabs.onAttached', () => {
+    it("Attaches a tab to another window and updates the store's state", () => {
+      const [callbackOnAttached] = mocks.chrome.tabs.onAttached.addListener.firstCall.args
```

Comment:
Let's just call this `callback` as we do throughout the file

Reply:
mock.restore() 
Great question. Yes, neglecting to reset/restore your mocks can cause issues. We’re using sinon-chrome which is a library someone made with a bunch of stubs representing the chrome API. You’ll see that in the teardown function we call chrome.reset() . That teardown function gets called after all tests run in whatever describe block the createMocks function got called in.

https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L7
https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L103
https://github.com/morehumaninternet/roar-extension/blob/54bc39645caaef5e0992900f8ac9ee190b9083ef/src/tests/integration/mocks.tsx#L113
If locally you comment out the chrome.reset function you’ll see things go haywire very fast.